### ISSUE TITLE
Inlining vs inout

### ISSUE THREAD

**Issue by therontarigo** [laurentlb/shader-minifier#471]
When inlining a function, Minifier behaves as though all parameters are qualified `inout`.

Contrived example reproducing the bug:

```GLSL
out vec4 O;
float i_func(float p) {
  return p-=2.,p*p;
}
void main() {
  float x=gl_FragCoord.x;
  O.x=i_func(x);
  O.y=x;
}
```

```GLSL
out vec4 v;
void main()
{
  float x=gl_FragCoord.x;
  v.x=(x-=2.,x*x);
  v.y=x;
}
```

With the current single-expression-only inlining, this hasn't been an issue in practice.  However it is likely to come up if allowing #470 .
The simplest fix is to enforce a rule:
Every parameter declaration of a function to be inlined must be either qualified `inout` or used immutably (no assignments).

**Comment by eldritchconundrum** [comment]
Forcing inlining can easily cause bugs, because it bypasses all the verifications that inlining the function will behave correctly, that are described here: https://github.com/laurentlb/shader-minifier/blob/3e25bdf74b9db57fe4a0328138b2cbca3fe303d9/src/inlining.fs#L172-L197

**Comment by therontarigo** [comment]
Ah, so perhaps this is user-error as I did not realize `i_` overrides this particular verification (I can't think of a legitimate reason for allowing it).  Sometimes `i_` is necessary on an already safe to inline function for overriding the heuristics to save some bytes and it would be helpful to not forgo all reasonable checks in that situation.

**Comment by therontarigo** [comment]
In a similar spirit to rule G requiring expression purity for `in` expressions, it would be reasonable to relax rule F to require that if `out` or `inout` is used, the argument at the call site must be a variable reference only.

**Comment by therontarigo** [comment]
> If the 'out' parameters are read from, inlining can change the behavior

Causing changes to already undefined behavior is acceptable.  But it can introduce a bug in the case that the variable used as the `out` argument is also used in an argument expression.

`int func (out int O, int I); `

If `func` reads O before writing it, it's UB - always.

safe: `(func(x,y), x+y)` `(x+y, func(x,y))`
prone to driver bugs: `(x+func(x,y))` `(func(x,y)+x)`
inlining can break the behavior: `(func(x,x))` `(func(x,x+y))`

**Comment by eldritchconundrum** [comment]
> Ah, so perhaps this is user-error as I did not realize `i_` overrides this particular verification (I can't think of a legitimate reason for allowing it). Sometimes `i_` is necessary on an already safe to inline function for overriding the heuristics to save some bytes and it would be helpful to not forgo all reasonable checks in that situation.

Yes, that's how I see it: detecting if it's safe or not to inline isn't easy, `i_` is given as an advanced (and dangerous) tool for overriding the minifier's analysis, so currently it's the user's fault for not making sure that overriding the minifier's analysis is safe. Maybe we could refine this (to selectively override only some rules?) but that seems complex for little benefits.

Improving the current rules is possible but it can be tricky to get the details right. They were chosen to balance multiple goals: preventing inlining from breaking the code, but still covering a lot of useful cases, but also not making the implementation too complex and the analysis too slow.

**Comment by therontarigo** [comment]
Would it make sense to have a way to override inlining heuristic (inlining hint) that's separate from the existing "force inline" feature?
